### ISSUE TITLE
Incorrect subdiagnostic suggestions for `unresolved-reference` diagnostics

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#584]
### Summary

https://github.com/astral-sh/ruff/pull/18444 added fantastic subdiagnostic suggestions for `unresolved-reference` diagnostics in methods. They work great for most cases, but for several edge cases they result in incorrect suggestions.

---

We should not add the subdiagnostic if the method is a staticmethod: the subdiagnostic does not make much sense here:

```py
class Foo:
    def __init__(self):
        self.x = 42

    @staticmethod
    def static_method():
        print(x)
```

Our diagnostic is currently:

```
error[unresolved-reference]: Name `x` used when not defined
  --> foo.py:9:15
   |
 7 |     @staticmethod
 8 |     def static_method():
 9 |         print(x)
   |               ^
10 |
11 |     @classmethod
   |
info: An attribute `x` is available: consider using `self.x`
info: rule `unresolved-reference` is enabled by default
```

---

We should not suggest attributes only available on instances if the method is a classmethod: the subdiagnostic does not make much sense here:

```py
class Foo:
    def __init__(self):
        self.x = 42

    @classmethod
    def class_method(cls):
        print(x)
```

Our diagnostic is currently:

```
error[unresolved-reference]: Name `x` used when not defined
 --> bar.py:7:15
  |
5 |     @classmethod
6 |     def class_method(cls):
7 |         print(x)
  |               ^
  |
info: An attribute `x` is available: consider using `self.x`
info: rule `unresolved-reference` is enabled by default
```

This should be fixed by using `SubclassOf::from(self.db(), class.default_specialization(self.db()))` [here](https://github.com/astral-sh/ruff/blob/8485dbb324212dab0e26d2afb5929097af129bbf/crates/ty_python_semantic/src/types/infer.rs#L6060) if it's a classmethod (but still using `Type::instance(self.db(), class.default_specialization(self.db()))` if it's an instance method).

---

The subdiagnostic message always says "consider using `self.<attribute>`". But we should inspect the function to check what the name of the first parameter is rather than assuming it's `self`. Even for instance methods, the name of the first parameter is often not `self` if it's an instance method on a metaclass, for example:

```py
class Foo:
    Y = 42

    @classmethod
    def class_method(cls):
        print(Y)


class Meta(type):
    Z = 42
    
    def instance_metaclass_method(cls):
        print(Z)
```

Our current diagnostics:

```
error[unresolved-reference]: Name `Y` used when not defined
 --> baz.py:6:15
  |
4 |     @classmethod
5 |     def class_method(cls):
6 |         print(Y)
  |               ^
  |
info: An attribute `Y` is available: consider using `self.Y`
info: rule `unresolved-reference` is enabled by default

error[unresolved-reference]: Name `Z` used when not defined
  --> baz.py:13:15
   |
12 |     def instance_metaclass_method(cls):
13 |         print(Z)
   |               ^
   |
info: An attribute `Z` is available: consider using `self.Z`
info: rule `unresolved-reference` is enabled by default
```

**Comment by MatthewMckee4** [comment]
Is there a plan to introduce KnownClass::Staticmethod? It seems like it may be useful for this

**Comment by AlexWaygood** [comment]
Huh, I assumed that we already had a `STATIC_METHOD` flag on this bitflag, but I guess not! I'd suggest adding it and detecting staticmethods in the same way that we detect classmethods

https://github.com/astral-sh/ruff/blob/5faf72a4d9b50c6e330165685e57fae14ca68b73/crates/ty_python_semantic/src/types/function.rs#L93

**Comment by MatthewMckee4** [comment]
Okay thanks, ill have a go at refactoring to use that

**Comment by MatthewMckee4** [comment]
I think detecting staticmethods still requires KnownClass::Staticmethod. I believe this is how we add the bitflag

```rs
Type::ClassLiteral(class) => {
    if class.is_known(self.db(), KnownClass::Classmethod) {
        function_decorators |= FunctionDecorators::CLASSMETHOD;
        continue;
    }
}
```

**Comment by AlexWaygood** [comment]
That's fine!

**Comment by MatthewMckee4** [comment]
I guess this can be added in another PR? Or should i add it here?

**Comment by carljm** [comment]
> I guess this can be added in another PR? Or should i add it here?

Seems fine to add it in the same PR, it's a simple addition following existing patterns exactly
